`decimal.Inexact` exception results in HTTP 500 errors for POST /jobs

Boto3's DynamoDB client requires numbers to have the `Decimal.decimal` data type, not the `float` data type.  We implement a conversion for parameters with floating point values at: https://github.com/ASFHyP3/hyp3/blob/master/apps/api/src/hyp3_api/util.py#L45

Some `float` values cannot be exactly converted to a `decimal` value.  `boto3.client('dynamodb').put_item` throws an exception for such values (traceback below)

We should update `convert_floats_to_decimals` to use rounding (or another strategy) when converting floats that cannot be exactly represented as decimals.

For more information:
- https://docs.python.org/3/library/decimal.html#decimal.Decimal.from_float
- https://docs.python.org/3/library/decimal.html#decimal.Context.create_decimal_from_float
- https://github.com/boto/boto3/issues/665

```
[ERROR]    2021-06-04T22:43:17.424Z    0cfe14c0-f836-4c1c-b08c-ccbdfc77ab32    Exception on /jobs [POST]
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/var/task/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/var/task/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/var/task/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/var/task/flask/_compat.py", line 39, in reraise
    raise value
  File "/var/task/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/connexion/decorators/decorator.py", line 48, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/security.py", line 327, in wrapper
    return function(request)
  File "/var/task/connexion/decorators/uri_parsing.py", line 144, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/validation.py", line 184, in wrapper
    response = function(request)
  File "/var/task/connexion/decorators/parameter.py", line 121, in wrapper
    return function(**kwargs)
  File "/var/task/hyp3_api/handlers.py", line 80, in post_jobs
    dynamo.put_jobs(jobs)
  File "/var/task/hyp3_api/dynamo.py", line 28, in put_jobs
    table.put_item(Item=item)
  File "/var/runtime/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/var/runtime/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
  File "/var/runtime/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/var/runtime/botocore/client.py", line 648, in _make_api_call
    request_dict = self._convert_to_request_dict(
  File "/var/runtime/botocore/client.py", line 694, in _convert_to_request_dict
    api_params = self._emit_api_params(
  File "/var/runtime/botocore/client.py", line 723, in _emit_api_params
    self.meta.events.emit(
  File "/var/runtime/botocore/hooks.py", line 356, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/var/runtime/botocore/hooks.py", line 228, in emit
    return self._emit(event_name, kwargs)
  File "/var/runtime/botocore/hooks.py", line 211, in _emit
    response = handler(**kwargs)
  File "/var/runtime/boto3/dynamodb/transform.py", line 195, in inject_attribute_value_input
    self._transformer.transform(
  File "/var/runtime/boto3/dynamodb/transform.py", line 253, in transform
    self._transform_parameters(
  File "/var/runtime/boto3/dynamodb/transform.py", line 260, in _transform_parameters
    getattr(self, '_transform_%s' % type_name)(
  File "/var/runtime/boto3/dynamodb/transform.py", line 274, in _transform_structure
    self._transform_parameters(
  File "/var/runtime/boto3/dynamodb/transform.py", line 260, in _transform_parameters
    getattr(self, '_transform_%s' % type_name)(
  File "/var/runtime/boto3/dynamodb/transform.py", line 285, in _transform_map
    params[key] = transformation(value)
  File "/var/runtime/boto3/dynamodb/types.py", line 107, in serialize
    return {dynamodb_type: serializer(value)}
  File "/var/runtime/boto3/dynamodb/types.py", line 234, in _serialize_m
    return dict([(k, self.serialize(v)) for k, v in value.items()])
  File "/var/runtime/boto3/dynamodb/types.py", line 234, in <listcomp>
    return dict([(k, self.serialize(v)) for k, v in value.items()])
  File "/var/runtime/boto3/dynamodb/types.py", line 107, in serialize
    return {dynamodb_type: serializer(value)}
  File "/var/runtime/boto3/dynamodb/types.py", line 208, in _serialize_n
    number = str(DYNAMODB_CONTEXT.create_decimal(value))
decimal.Inexact: [<class 'decimal.Inexact'>, <class 'decimal.Rounded'>]
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+
+## [1.1.7]
+### Fixed
+- POST `/jobs` no longer throws 500 for`decimal.Inexact` errors
+
 ## [1.1.6]
 ### Changed
 - Landsat and Sentinel-2 autoRIFT jobs will utilize 1/3 less memory

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [1.1.7]
 ### Fixed
-- POST `/jobs` no longer throws 500 for`decimal.Inexact` errors
+- POST `/jobs` no longer throws 500 for`decimal.Inexact` errors, fixes [#444](https://github.com/ASFHyP3/hyp3/issues/444)
 
 ## [1.1.6]
 ### Changed

--- a/apps/api/src/hyp3_api/util.py
+++ b/apps/api/src/hyp3_api/util.py
@@ -42,7 +42,7 @@ def get_job_count_for_month(user):
 
 def convert_floats_to_decimals(element):
     if type(element) is float:
-        return Decimal(element)
+        return Decimal(str(element))
     if type(element) is list:
         return [convert_floats_to_decimals(item) for item in element]
     if type(element) is dict:

--- a/tests/api/test_dynamo.py
+++ b/tests/api/test_dynamo.py
@@ -1,6 +1,6 @@
 from api.conftest import list_have_same_elements
 
-from hyp3_api import dynamo
+from hyp3_api import dynamo, util
 
 
 def test_count_jobs(tables):
@@ -357,6 +357,43 @@ def test_query_jobs_sort_order(tables):
     ]
     for item in [table_items[2], table_items[0], table_items[1]]:
         tables['jobs_table'].put_item(Item=item)
+
+    response, _ = dynamo.query_jobs('user1')
+    assert response == table_items
+
+
+def test_decimal_conversion(tables):
+    table_items = [
+            {
+                'job_id': 'job1',
+                'job_type': 'RTC_GAMMA',
+                'name': 'name1',
+                'user_id': 'user1',
+                'status_code': 'status1',
+                'request_time': '2000-01-03T00:00:00+00:00',
+                'float_value': 30.04,
+            },
+            {
+                'job_id': 'job2',
+                'job_type': 'RTC_GAMMA',
+                'name': 'name1',
+                'user_id': 'user1',
+                'status_code': 'status1',
+                'request_time': '2000-01-02T00:00:00+00:00',
+                'float_value': 0.0,
+            },
+            {
+                'job_id': 'job3',
+                'job_type': 'INSAR_GAMMA',
+                'name': 'name2',
+                'user_id': 'user1',
+                'status_code': 'status1',
+                'request_time': '2000-01-01T00:00:00+00:00',
+                'float_value': 0.1,
+            },
+        ]
+    for item in table_items:
+        tables['jobs_table'].put_item(Item=util.convert_floats_to_decimals(item))
 
     response, _ = dynamo.query_jobs('user1')
     assert response == table_items

--- a/tests/api/test_dynamo.py
+++ b/tests/api/test_dynamo.py
@@ -1,3 +1,5 @@
+from decimal import Decimal
+
 from api.conftest import list_have_same_elements
 
 from hyp3_api import dynamo, util
@@ -396,4 +398,6 @@ def test_decimal_conversion(tables):
         tables['jobs_table'].put_item(Item=util.convert_floats_to_decimals(item))
 
     response, _ = dynamo.query_jobs('user1')
-    assert response == table_items
+    assert response[0]['float_value'] == Decimal('30.04')
+    assert response[1]['float_value'] == Decimal('0.0')
+    assert response[2]['float_value'] == Decimal('0.1')

--- a/tests/api/test_dynamo.py
+++ b/tests/api/test_dynamo.py
@@ -366,34 +366,34 @@ def test_query_jobs_sort_order(tables):
 
 def test_decimal_conversion(tables):
     table_items = [
-            {
-                'job_id': 'job1',
-                'job_type': 'RTC_GAMMA',
-                'name': 'name1',
-                'user_id': 'user1',
-                'status_code': 'status1',
-                'request_time': '2000-01-03T00:00:00+00:00',
-                'float_value': 30.04,
-            },
-            {
-                'job_id': 'job2',
-                'job_type': 'RTC_GAMMA',
-                'name': 'name1',
-                'user_id': 'user1',
-                'status_code': 'status1',
-                'request_time': '2000-01-02T00:00:00+00:00',
-                'float_value': 0.0,
-            },
-            {
-                'job_id': 'job3',
-                'job_type': 'INSAR_GAMMA',
-                'name': 'name2',
-                'user_id': 'user1',
-                'status_code': 'status1',
-                'request_time': '2000-01-01T00:00:00+00:00',
-                'float_value': 0.1,
-            },
-        ]
+        {
+            'job_id': 'job1',
+            'job_type': 'RTC_GAMMA',
+            'name': 'name1',
+            'user_id': 'user1',
+            'status_code': 'status1',
+            'request_time': '2000-01-03T00:00:00+00:00',
+            'float_value': 30.04,
+        },
+        {
+            'job_id': 'job2',
+            'job_type': 'RTC_GAMMA',
+            'name': 'name1',
+            'user_id': 'user1',
+            'status_code': 'status1',
+            'request_time': '2000-01-02T00:00:00+00:00',
+            'float_value': 0.0,
+        },
+        {
+            'job_id': 'job3',
+            'job_type': 'INSAR_GAMMA',
+            'name': 'name2',
+            'user_id': 'user1',
+            'status_code': 'status1',
+            'request_time': '2000-01-01T00:00:00+00:00',
+            'float_value': 0.1,
+        },
+    ]
     for item in table_items:
         tables['jobs_table'].put_item(Item=util.convert_floats_to_decimals(item))
 

